src/librustc/infer/canonical/canonicalizer.rs

```diff
@@ -295,7 +304,8 @@ impl<'cx, 'gcx, 'tcx> Canonicalizer<'cx, 'gcx, 'tcx> {
         infcx: Option<&'cx InferCtxt<'cx, 'gcx, 'tcx>>,
         tcx: TyCtxt<'cx, 'gcx, 'tcx>,
         canonicalize_region_mode: CanonicalizeRegionMode,
-    ) -> (Canonicalized<'gcx, V>, CanonicalVarValues<'tcx>)
+        var_values: &'cx mut SmallCanonicalVarValues<'tcx>
+    ) -> Canonicalized<'gcx, V>
```

Comment:
Why not just return this? I guess it's more efficient this way...?

Reply:
Yes. Copying the `SmallCanonicalVars` reduces the size of the win by about 20--25%. I figure we need every saving we can get for NLL!